5.4 categorical color (#75)
UBC-MDS/exploratory-data-viz: chapters/en/slides/module5/module5_04.Rmd

to `#ffffff` for white ("full color")
and [are easiest chosen via color pickers such as this one](https://colorpicker.me/).

---

## Don't use more than 5-8 distinct hues for categories

<center><img src=/module5/hue-guidelines.png></img></center>

Notes:
Although many of the categorical color schemes contain ten or more hues,
if is often not a good idea to use all that many.

The guidelines on what is too many hues differ between different sources
and also depends on your use case.
A good rule of thumb is that when you get to around five different hues,
you should really consider if this is the best way to represent your data
or if you could split it up into multiple visualizations instead.

If the data is neatly organized in well-separated clusters,
it is possible that you could visualize more than five colors
(maybe even ten) effectively,
but in data where the datapoints are more mixed,
you will rarely, if ever, be able to go this high.

---

## Too many hues are impossible to distinguish

```{python}
alt.Chart(cars).mark_point(size=70, filled=True).encode(
    x='Horsepower',
    y='Miles_per_Gallon',
    color=alt.Color('Name', title=None))

> ```

Author: ![image](https://user-images.githubusercontent.com/4560057/108153442-07006500-7090-11eb-989f-69f72791a740.png)

---
params:
  dynamictitle: "module5_04"
title: "`r params$dynamictitle`"
output:
  md_document:
    variant: gfm
    pandoc_args: "--markdown-headings=atx"
---

```{r setup, include=FALSE}
source('../../../../slide-setup.R')
# Remember to also manually update the YAML title above
```

type: slides

> # Effective use of color for categorical data

Reviewer: ```suggestion
# Effective use of colour for categorical data
```

Notes:

> In this slide deck you will learn how to select colors that accurately represents your data

Reviewer: ```suggestion
In this slide deck you will learn how to select colours that accurately represents your data
```

and effectively communicate it to the reader.

---

> ## Color choices can have profound impact on the interpretation of the results

Reviewer: ```suggestion
## Colour choices can have profound impact on the interpretation of the results
```

![](/module5/color-importance-vessels.png)

Notes:

> In module 2 we discussed how changing the colorscheme drastically improved a clinical outcome

Reviewer: ```suggestion
In module 2 we discussed how changing the colourscheme drastically improved a clinical outcome
```

by almost doubling physician's detection rate of blodd vessles regions indicative of potential future heart disease 
(from 40% to 70%).

In this slide deck we will discuss some of the consideration

> that goes into evaluating if a colorscheme (also called a colormap)

Reviewer: ```suggestion
that goes into evaluating if a colourscheme (also called a colourmap)
```

is suitable for the data you are visualizing.

---

> ## Color can be broken down into hue, lightness, and saturation

Reviewer: ```suggestion
## Colour can be broken down into hue, lightness, and saturation
```

<center><img src="/module5/hsl-cylinder.png" width="100%"></img></center>

Notes:

> There are several different ways to represent colors.
> You might arleady have heard of RGB (red, green, blue) or CMYK (cyan, magenta, yellow, black)
> where different amount of a set of base colors are combined to create all possible colors.
> 
> When discussing colors in the context of data visualization,

Reviewer: ```suggestion
There are several different ways to represent colours.
You might arleady have heard of RGB (red, green, blue) or CMYK (cyan, magenta, yellow, black)
where different amount of a set of base colours are combined to create all possible colours.

When discussing colours in the context of data visualization,
```

> We will be describing them with the parameters "hue", "saturation" and "lightness".

Author: good catch ,I used the wrong image

The left schematic in this slide is a 2D circular representation including only hue and saturation.
When we add lightness this circle grows into cylinder with lightness as the height.

On the next slide we will break down exactly what each of these three parameters represent.

---

> ## Hue, lightness, and saturation describes different properties of colors

Reviewer: ```suggestion
## Hue, lightness, and saturation describes different properties of colours
```

<center><img src="/module5/hsl-questions.png" width="70%"></img></center>

Notes:
We could verbalize the changes we saw in the previous slide in the following manner:

> Hue is what we traditionally think of as the “color”, is it red, blue, etc?

Reviewer: ```suggestion
Hue is what we traditionally think of as the “colour”, is it red, blue, etc?
```

Each hue can have a varying saturation, which ranges from a dull, greyish appearance to a vibrant fully saturated hue.

> Lightness is how bright the color is. For every hue, it starts at black (no lightness) and ends at white (full lightness).

Reviewer: ```suggestion
Lightness is how bright the colour is. For every hue, it starts at black (no lightness) and ends at white (full lightness).
```

Hue and lightness are the most important from a data visualization perspective,
and we use them for categorical and quantitative data, respectively. 

Saturation is often used more as a stylistic choice

> to decide whether the colors we use should be muted/desaturated or vivid/saturated.

Reviewer: ```suggestion
to decide whether the colours we use should be muted/desaturated or vivid/saturated.
```

---

## Hues are used to distinguish categorical values

```{python}
# This code cell should be hidden but echo=False does not work with the knitr output hook
import altair as alt
from vega_datasets import data

cars = data.cars()
chart = alt.Chart(cars).mark_point(size=50, filled=True).encode(
    x='Horsepower',
    y='Miles_per_Gallon',
    color=alt.Color('Origin', title=None))

> (chart | chart.encode(alt.Color('Origin', scale=alt.Scale(scheme='blues'), title=None))).resolve_scale(color='independent')

Reviewer: Can altair use Canadian spelling like ggplot2 can?

Author: Nope

> (chart | chart.encode(alt.Color('Origin', scale=alt.Scale(scheme='blues'), title=None))).resolve_scale(color='independent')

Author: This is a hidden cell so students won't see it

```

Notes:
Hue is useful to distinguish between categories,
because it is often relatively easy for us to to say that different hues are distinct.
We can see that blue, orange, red, etc are different and don’t easily mix them up.

The leftmost plot uses different hues
to separate the points belonging to different categories.

> We can quickly identify that there is a tota of three distinct colors being used here

Reviewer: ```suggestion
We can quickly identify that there is a total of three distinct colours being used here
```

even without looking at the legend.

In the plot to the right
we used lightness within a single hue (blue) to label the different categories.
Here it is much harder to say how many different categories there are,
and even when we have the legend,
it is difficult to tell which points are which shade of blue.

---

> ## It is often better to use established color schemes instead of making your own

Reviewer: ```suggestion
## It is often better to use established colour schemes instead of making your own
```

<center><img src=/module5/color-schemes.png></img></center>

Notes:

> There are many different hues we could use to represent our colors

Reviewer: ```suggestion
There are many different hues we could use to represent our colours
```

and some might not go as well together as others,
so how do we decide which ones to use?

Fortunately,

> we don’t have to design our own combination of colors to use,
> but can pick from combinations designed by experts to be easy to tell apart
> and in most cases also suitable for people with color vision deficiencies.
> 
> These color combinations are referred to as color schemes in Altair,
> but you might also hear them being called colormaps or color palettes.
> 
> In this slide, you can see some of the color schemes for categorical values
> that are built into Altair.
> 
> The default one is “tableau10”
> and it was named this way because it was originally designed by the company Tableau,
> but is now one of the most common categorical color schemes used in data visualization.
> [All Altair color schemes can be viewed here](https://vega.github.io/vega/docs/schemes/).

Reviewer: ```suggestion
we don’t have to design our own combination of colours to use,
but can pick from combinations designed by experts to be easy to tell apart
and in most cases also suitable for people with colour vision deficiencies.

These color combinations are referred to as colour schemes in Altair,
but you might also hear them being called colormaps or colour palettes.

In this slide, you can see some of the colour schemes for categorical values
that are built into Altair.

The default one is “tableau10”
and it was named this way because it was originally designed by the company Tableau,
but is now one of the most common categorical colour schemes used in data visualization.
[All Altair colour schemes can be viewed here](https://vega.github.io/vega/docs/schemes/).
```

---

> ## Specifying color schemes in Altair

Reviewer: ```suggestion
## Specifying colour schemes in Altair
```

```{python}
alt.Chart(cars).mark_point(size=70, filled=True).encode(
    x='Horsepower',
    y='Miles_per_Gallon',
    color=alt.Color('Origin', title=None, scale=alt.Scale(scheme='set1')))
```

Notes:

> You can change the color scheme of a plot by specifying its name as a string

Reviewer: ```suggestion
You can change the colour scheme of a plot by specifying its name as a string
```

to the `scheme` parameter inside `alt.Scale`.

> To easier see the changes we are making to the colors,

Reviewer: ```suggestion
To more easiely see the changes we are making to the colours,
```

Reviewer: This is the last one I am going to fix, I am 99% sure we discussed that we were going to use Canadian spelling, so if I am correct, please continue to change this going forward. You might have to change it retroactively in this module (and others) as I was not paying attention to this earlier.

Author: yup, I will do find an replace on all afterwards.

we also increase the size of all points 
via the `size` parameter of the mark.

---

## Redundant coding can make charts easier to interpret

```{python}
alt.Chart(cars).mark_point(size=70, filled=True).encode(
    x='Horsepower',
    y='Miles_per_Gallon',
    color=alt.Color('Origin', title=None, scale=alt.Scale(scheme='set1')),
    shape='Origin')
```

Notes:
Although many of the color schemes in Altair 
are designed according to the guidelines for effective color uses,
they can sometimes be difficult to interpret 
for people with color vision deficiences.

> Esppecially color schemes that mix red and green

Reviewer: ```suggestion
Especially colour schemes that mix red and green
```

as the one in the previous slide.

In addition to using a more suitable color scheme (such as the default),
we could also change the shape of the points for each category.

Although this is technically redundant
since the color is already used for the categorical groups,

> it can make your visualization more effective.

Reviewer: Why? Because the marks "pop out" as being different, and this is helpful for folks who do not have a disability as well as those who do.

For line plots we could achieve a similar effect via the `strokeDash` encoding
instead of `shape`,
which together with direct labeling from the previous slide deck

> can facilitate interpretation of these plots.

Reviewer: Nice!

---

## Specifying custom colors

```{python}
colors = ['coral', '#4682b4', 'rebeccapurple']
alt.Chart(cars).mark_point(size=70, filled=True).encode(
    x='Horsepower',
    y='Miles_per_Gallon',
    color=alt.Color('Origin', title=None, scale=alt.Scale(range=colors)))

> ```

Reviewer: Can we add `shape='Origin'` here now that they know about it?

Reviewer: And can we please have nicely formatted axes labels (this is our effective plots module...). Can you fix this on the previous plots as well?

You can specify colors either by their HTML/CSS name (such as 'coral')
or their hex code (such as '#4682b4').
[All HTML/CSS color names can be found in the image in this post](https://stackoverflow.com/a/37232760/2166823).

Hex codes are defines over the range `#000000` for black ("zero color")
to `#ffffff` for white ("full color")
and [are easiest chosen via color pickers such as this one](https://colorpicker.me/).

---

## Don't use more than 5-8 distinct hues for categories

<center><img src=/module5/hue-guidelines.png></img></center>

Notes:
Although many of the categorical color schemes contain ten or more hues,
if is often not a good idea to use all that many.

The guidelines on what is too many hues differ between different sources
and also depends on your use case.
A good rule of thumb is that when you get to around five different hues,
you should really consider if this is the best way to represent your data
or if you could split it up into multiple visualizations instead.

If the data is neatly organized in well-separated clusters,
it is possible that you could visualize more than five colors
(maybe even ten) effectively,
but in data where the datapoints are more mixed,

> you will rarely, if ever, be able to go this high.

Reviewer: You should explain why here - because we start to become unable to distinguish them! You hint at this, but do not explicitly say this. You should!

---

## Too many hues are impossible to distinguish

```{python}
alt.Chart(cars).mark_point(size=70, filled=True).encode(
    x='Horsepower',
    y='Miles_per_Gallon',
    color=alt.Color('Name', title=None))
```

Notes:
This is an example of what happens when too many categorical hues are used.
In this plot,
we can not separate the colors
even if we spent a considerable amount of effort studying the chart.

As we saw on the last slide,
categorical color schemes have a limited amount of hues
so in addition of it being hard to differentiate this many hues in general,
it become practically impossible in cases
where there are more categories than coor hues,
since the color scheme starts repeating as in the chart.

A better approach here would have been to label specific points of interest directly

> and keeping the rest as either a single color or using a categorical variable with fewer values, e.g. by grouping the cars into brands rather than their full model name.

Reviewer: Can we add a slide showing this better way of visualizing this data? I think the learners would really benefit from seeing this, over having a verbal description.

```

Notes:
A few slides ago we saw how encoding the same variable as both shape and color
could make our figures more effective.

Quite frequently,
you might come across bar charts where the author of the visualization
has added a distinct color for each bar.

This can work when there are relatively few bars,
but it usually does not add anything
and can become directly disorienting when there are many bars
as in this slide.

It is usually preferable to color all the bars in the same color
(or after another categorical variable, such as "Origin" in this data),
and let the axis labels alone identify which category the bar belongs too.

One exception to this is if each axis label
is broken down into subcategories
where each location has multiple bars based on another categorical variable,
then color can help.

---

## Use consistent coloring between subplots even when it is redundant

```{python}
chart = alt.Chart(cars).mark_bar().encode(
    alt.X('Origin', title=None),
    alt.Y('mean(Horsepower)'),
    alt.Color('Origin'))
alt.hconcat(chart, chart.mark_line().encode(x='Year', strokeDash='Origin'))

> ```

Reviewer: The second legend here has no colour, can we fix that? Also, do we want origin twice here?

Author: There doesn't seem to be an easy why to combine with multiple charts, for a single chart it works as desired.
